### PR TITLE
Configure ALPN for callback scenarios 

### LINKED ISSUE #31303: Make ALPN configuration easier for use with TlsHandshakeCallbackOptions
> Thanks for adding this. We're using `ServerOptionsSelectionCallback` and `SslStreamCertificateContext` to enable hosting thousands of domains and selecting the right ssl certificates with Kestrel. Though `X509Certificate2` on Windows does have temporary file creation issues that don't disappear with dispose that we have to deal with.
> 
> I had to spend some time to figure this out and to get it working with HTTP/2. Posting some sample code here for the benefit of others trying to use `ServerOptionsSelectionCallback`, for inclusion in documentation, and if anything with the sample code triggers changes/improvements.
> 
> We needed `ConfigureAlpn` to enable HTTP/2. Maybe this method should be made public in `HttpsConnectionMiddleware` or called internally in `HttpsConnectionMiddleware` for us?
> 
> Also would be great to have access to `ConnectionContext` so we can get the connecting ip, listening port, etc. to make certain decisions at this stage or to abort the connection if we didn't find a matching certificate or for any other reasons. If there is a best practice there on how to drop the connection when a matching certificate is not found, would be glad to hear it. Right now, we just return null as the certificate and let Kestrel drop the connection.
> 
> ```c#
> listenOptions.UseHttps(ServerOptionsSelectionCallback, state: listenOptions.Protocols);
> 
> ValueTask<SslServerAuthenticationOptions> ServerOptionsSelectionCallback(
> 	SslStream stream,
> 	SslClientHelloInfo clientHelloInfo,
> 	object state,
> 	CancellationToken cancellationToken)
> {
> 	var serverOptions = new SslServerAuthenticationOptions
> 	{
> 		EnabledSslProtocols = SslProtocols.Tls12 | SslProtocols.Tls13,
> 		CertificateRevocationCheckMode = X509RevocationMode.NoCheck,
> 		ServerCertificateContext = SslStreamCertificateContext.Create(certificate, intermediateCertificates, offline: true)
> 	};
> 
> 	//ConfigureAlpn / Enable Http2
> 	var httpProtocols = (HttpProtocols)state;
> 	ConfigureAlpn(serverOptions, httpProtocols);
> 
> 	return new ValueTask<SslServerAuthenticationOptions>(serverOptions);
> }
> 
> //Copy of method from:
> //https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs
> void ConfigureAlpn(SslServerAuthenticationOptions serverOptions, HttpProtocols httpProtocols)
> {
> 	serverOptions.ApplicationProtocols = new List<SslApplicationProtocol>();
> 
> 	// This is order sensitive
> 	if ((httpProtocols & HttpProtocols.Http2) != 0)
> 	{
> 		serverOptions.ApplicationProtocols.Add(SslApplicationProtocol.Http2);
> 		// https://tools.ietf.org/html/rfc7540#section-9.2.1
> 		serverOptions.AllowRenegotiation = false;
> 	}
> 
> 	if ((httpProtocols & HttpProtocols.Http1) != 0)
> 	{
> 		serverOptions.ApplicationProtocols.Add(SslApplicationProtocol.Http11);
> 	}
> }
> ```

Thanks @nitinag for reporting this issue in https://github.com/dotnet/aspnetcore/pull/25390#issuecomment-807328841.

I think having a public static `ConfigureAlpn` method like @nitinag suggests is a reasonable approach. It's internal rather than private today so [SniOptionsSelector can call it](https://github.com/dotnet/aspnetcore/blob/ec0841d8d7e0c2e700c52bd4d89e7687d9ed78a1/src/Servers/Kestrel/Core/src/Internal/SniOptionsSelector.cs#L95) which probably should have been a sign it should be made public in the first place.

My only concern is that people still won't know to call it and therefore unintentionally disable HTTP/2 when trying to write custom SNI logic. API docs can help here, but not everyone reads it.

I wonder if instead we should automatically populate `ApplicationProtocols` if its null. The question there becomes what if I want to disable ALPN? Is that important? Could that be done with an empty list instead of a null list?

We shouldn't let perfect be the enemy of the good though. If we cannot autopupulate the ALPN list for whatever reason, we should just expose `ConfigureAlpn`. It shouldn't be necessary for people to copy/paste this code.

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #31303 by stashing HttpProtocols on TlsHandshakeCallbackOptions just like we did for HttpsConnectionAdapterOptions. If either of the callbacks don't supply ALPN settings then we will. ApplicationProtocols can be set to an empty list to disable ALPN.

Is this too magical?

Is it breaking for people that were using ServerOptionsSelectionCallback and suddenly HTTP/2 starts working?

### CODE REVIEW

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs [hunk 1]
```diff
@@ -793,6 +793,65 @@ void ConfigureListenOptions(ListenOptions listenOptions)
...
+        [ConditionalFact]
+        [OSSkipCondition(OperatingSystems.MacOSX, SkipReason = "ALPN not supported")]
+        public async Task ServerOptionsSelectionCallback_SetsALPN()
+        {
+            static void ConfigureListenOptions(ListenOptions listenOptions)
+            {
+                listenOptions.UseHttps((_, _, _, _) =>
+                    ValueTask.FromResult(new SslServerAuthenticationOptions()
+                    {
+                        ServerCertificate = _x509Certificate2,
```
Comment (halter73) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs:805]: Can we add a test case confirming that adding

```suggestion
                        ServerCertificate = _x509Certificate2,
                        ApplicationProtocols = new()
```

disable ALPN? Assuming it does ofc.

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs (hunk 1 above)
Comment (Tratcher) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs:805]: It worked on Windows but fails on Ubuntu 😢. I'll file a runtime bug for that.

Comment (Tratcher) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs:805]: Fixed in rc1 (main), https://github.com/dotnet/runtime/pull/55772. Waiting for a runtime update.

**Pull request merged**